Userland/Libraries/LibJS/Runtime/AbstractOperations.cpp

```diff
@@ -544,10 +544,19 @@ Object* create_mapped_arguments_object(GlobalObject& global_object, FunctionObje
 }
 
 // 7.1.21 CanonicalNumericIndexString ( argument ), https://tc39.es/ecma262/#sec-canonicalnumericindexstring
-Value canonical_numeric_index_string(GlobalObject& global_object, Value argument)
+Value canonical_numeric_index_string(GlobalObject& global_object, PropertyName const& property_name)
 {
+    // NOTE: If the property name is a number type (An implementation-defined optimized
+    // property key type), it can be treated as a string property that has already been
+    // converted successfully into a canonical numeric index.
+
+    VERIFY(property_name.is_string() || property_name.is_number());
+
+    if (property_name.is_number())
+        return Value(property_name.as_number());
+
     // 1. Assert: Type(argument) is String.
-    VERIFY(argument.is_string());
+    Value argument = js_string(global_object.vm(), property_name.as_string());
```

Comment:
Earlier note about `auto` applies :)

```suggestion
    auto argument = js_string(global_object.vm(), property_name.as_string());
```

Reply:
`auto` here gives the wrong type since `js_string` gives PrimitiveString.
I wrapped it like `Value(js_string(...))`. Is that preferred or should it be Value on the left?

Reply:
ah, right - this is much better than an implicit cast based on the value type though, so :+1: